experimental/devices/inky/inky.go

```diff
@@ -0,0 +1,291 @@
+	"time"
+
+	"periph.io/x/periph/conn"
+	"periph.io/x/periph/conn/display"
+	"periph.io/x/periph/conn/gpio"
+	"periph.io/x/periph/conn/physic"
+	"periph.io/x/periph/conn/spi"
+)
+
+const (
+	// Constants for an Inky pHAT
+	cols = 104
+	rows = 212
+
+	speed      = 488 * physic.KiloHertz
+	spiBits    = 8
+	chunkSize  = 4096
+	spiCommand = gpio.Low
+	spiData    = gpio.High
+)
+
+type Color int
+
+const (
+	Black  = iota
+	Red    = iota
+	Yellow = iota
+	White  = iota
+)
+
+var borderColor = map[Color]byte{
+	Black:  0x00,
+	Red:    0x33,
+	Yellow: 0x33,
+	White:  0xff,
+}
+
+// NewpHAT opens a handle to an Inky pHAT.
+func NewpHAT(p spi.Port, dc gpio.PinOut, reset gpio.PinOut, busy gpio.PinIn, color Color) (*Dev, error) {
+	if color != Black && color != Red && color != Yellow {
+		return nil, fmt.Errorf("Unsupported color: %v", color)
+	}
+
+	c, err := p.Connect(speed, spi.Mode0, spiBits)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to inky over spi: %v", err)
+	}
+
+	d := &Dev{
+		c:      c,
+		dc:     dc,
+		r:      reset,
+		busy:   busy,
+		color:  color,
+		border: Black,
+	}
+
+	return d, nil
+}
+
+// Dev is a handle to an Inky.
+type Dev struct {
+	c conn.Conn
+	// Low when sending a command, high when sending data.
+	dc gpio.PinOut
+	// Reset pin, active low.
+	r gpio.PinOut
+	// High when device is busy.
+	busy gpio.PinIn
+
+	// Color of device screen (red, yellow or black).
+	color Color
+	// Modifiable color of border.
+	border Color
+}
+
+// SetBorder changes the border color. This will not take effect until the next Draw().
+func (d *Dev) SetBorder(c Color) {
+	d.border = c
+}
+
+// String implements conn.Resource.
+func (d *Dev) String() string {
+	return "InkyPHat"
+}
+
+// Halt implements conn.Resource
+func (d *Dev) Halt() error {
+	return nil
+}
+
+// ColorModel implements display.Drawer
+// Maps white to white, black to black and anything else as red. Red is used as
+// a placeholder for the display's third color, i.e., red or yellow.
+func (d *Dev) ColorModel() color.Model {
+	return color.ModelFunc(func(c color.Color) color.Color {
+		r, g, b, _ := c.RGBA()
+		if r == 0 && g == 0 && b == 0 {
+			return color.RGBA{
+				R: 0,
+				G: 0,
+				B: 0,
+				A: 255,
+			}
+		} else if r == 0xffff && g == 0xffff && b == 0xffff {
+			return color.RGBA{
+				R: 255,
+				G: 255,
+				B: 255,
+				A: 255,
+			}
+		}
+		return color.RGBA{
+			R: 255,
+			G: 0,
+			B: 0,
+			A: 255,
+		}
+	})
+}
+
+// Bounds implements display.Drawer
+func (d *Dev) Bounds() image.Rectangle {
+	return image.Rect(0, 0, rows, cols)
+}
+
+// Draw implements display.Drawer
+func (d *Dev) Draw(dstRect image.Rectangle, src image.Image, srcPtrs image.Point) error {
+	if dstRect != d.Bounds() {
+		return fmt.Errorf("Partial update not supported")
+	}
+
+	if src.Bounds() != d.Bounds() {
+		return fmt.Errorf("Image must be the same size as bounds: %v", d.Bounds())
+	}
+
+	b := src.Bounds()
+	// Black/white pixels.
+	white := make([]bool, rows*cols)
```

Comment:
Sadly, []bool in Go is not optimized, so it's using 1 byte per bit, so a 8x blow up. It's not a show stopper here but would be nice to fix later. There is already an example of bitmap image format at https://periph.io/x/periph/devices/ssd1306/image1bit but it's not the right format.

You're not required to do this but one possibility would be to recreate a 2 bit image format that would be able to represent the 3 colors. Technically you could pack it to 1.5 bits per pixel, but that'd be more work than worth it. You already started with ColorModel() above. :)

Reply:
I thought about it but decided the complexity wasn't worth it for now - the memory blowup is only ~350KB and I think the minimum RAM you can get on a Pi is 512MB these days. Might be worth it for a wHAT though (~2MB).